### ISSUE TITLE
maxRows does not work

### ISSUE THREAD

**Issue by giulianok** [STRML/react-grid-layout#1278]
I'm trying to build a Grid 2x6 where some of the items are `static`. The problem I'm having is that I don't want to add more rows, I want to keep only 3 rows and re-order the Grid

Here's a demo where the picture goes to the end of the grid, making a new row, instead of moving to available spot

![Sep-04-2020 13-46-57](https://user-images.githubusercontent.com/8062245/92288337-a236c100-eeda-11ea-9cbd-ae31377fc99e.gif)

Also, if I want to move it to the end and make a new row I can as well (instead i would like to block it and not allow that)

![Screen Shot 2020-09-04 at 4 50 36 PM (2)](https://user-images.githubusercontent.com/8062245/92288418-dca05e00-eeda-11ea-8a6a-9dc45c794a91.png)
![Screen Shot 2020-09-04 at 4 50 38 PM (2)](https://user-images.githubusercontent.com/8062245/92288424-de6a2180-eeda-11ea-936e-2e66d7e1bc2c.png)
![Screen Shot 2020-09-04 at 4 50 41 PM (2)](https://user-images.githubusercontent.com/8062245/92288426-e033e500-eeda-11ea-993d-39509cb90ac2.png)

I moved my code a playground as well https://codesandbox.io/s/compassionate-rain-p2074?file=/src/ShowcaseLayout.js

**Comment by manang** [comment]
Hi, did you solve the issue?
Thanks.

**Comment by giulianok** [comment]
not @manang, I just focused on other things I have to do and I will get back to this feature soon. Looks like nobody has a quick answer for it, that tells me I need to go deep and investigate 🤓

**Comment by manang** [comment]
there is this very old pr that  can help:
https://github.com/STRML/react-grid-layout/pull/657/commits/ff2e60b1e2d50c8042fbadfce0dad7b996795d7a

**Comment by lanesawyer** [comment]
Could we finally get a fix on this? [There's a PR read to go for it that's five years old.](https://github.com/react-grid-layout/react-grid-layout/pull/657)

I'm unable to use the draggable functionality because the minimum rows are completely broken, and I'm frustrated that it's even a documented, available prop in the first place. Highly misleading.

**Comment by STRML** [comment]
It's *not* a documented, available prop. I have no idea where you are getting that from. It is only in the source with the specific callout that it doesn't work as expected.

You are free to consider other libraries. You are not paying me, you are simply requesting free work.

Want to contribute? Rebase that PR on master and add tests so we can be sure that it works and will continue to work. That would be very helpful.

**Comment by lanesawyer** [comment]
My bad, it came up in the types so I assumed it was a feature.

**Comment by lanesawyer** [comment]
Since I was being a dick, I'll be helpful to others with a workaround:

This advice depends on the use case, but if you're not wanting items to get pushed down into an extra row you didn't want to exist and instead have the items move horizontally, you can get that behavior by switching `compactType` to `horizontal`.

Play with `compactType` in combination with your `cols` and layout settings and you won't have as much of a need for `maxRows`.

**Comment by WarrenGreen** [comment]
@lanesawer That does the trick!

**Comment by prechapl** [comment]
Switching compactType to horizontal does not prevent an extra row being added for me. Specifically when a grid item is dragged downward a new row is created and upper grid spots are left blank.  Has anyone found a solution for this issue?

**Comment by AriEncarnacion** [comment]
@prechapl As far as I know the requests and issues with a `maxRows` feature has been discussed back and forth but no library solution has been implemented yet. 

We met a similar use case where we needed the following:
1. Users can use the "drag & drop" functionality as seen here to add new Nodes (what we're calling the boxes in the grid)
2. Users can use all features of RGL; resize, arrange, and delete Nodes
3. Users are restricted to a specific grid size ( for example, 16 cols and 9 rows)

[3] causes an issue as there's no available prop to restrict users from adding new rows. We implemented workarounds for the above requirements as such:

At all times:
We save the Layout and current grid size

Adding Nodes:
When the user drags from outside and drops into the grid, if dropped outside of the current grid size, the Node is first dropped, but we remove it from the Layout and throw an error to the user, "Node dropped outdside grid dimensions" and the RGL naturally snaps back to its designated size.

Resizing Nodes:
When resizing Nodes, we use similar functionality but check the sum of the size and position of the Node. If it exceeds the grid size, it snaps back to its previous size.

Arranging Nodes:
This is the most difficult to handle. RGL's natural behavior is to add rows if the compaction or arrangement of Nodes meet certain requirements. To address this, we constantly check to see if nodes have been pushed outside of the grid's required dimensions. If they do, they're pushed to the next "available" spot. If no spaces are available, we throw an error to the User and the Layout snaps back to its previous arrangement.

With these workarounds, adding and resizing with fixed rows and columns works nicely, but arrangement can be buggy. We're still looking into some further workarounds and also looking at the RGL library to see if we can open a PR for an actual fix. I can't give a timeline on this though.

**Comment by madscientist111** [comment]
Hey @AriEncarnacion, is your solution available anywhere to look up?

We've been trying a lot of items on https://github.com/parth-chovatiya/react-grid-layout/tree/master-new

The time we have given this thing to fix is much more than I'd like to admit. So any kind of help will be very much appreciated.

We are trying out similar to what you have done and yes, the arranging node is the trickiest.

Some of the things we have tried::
1. We realised that everything comes to compact() in the end - so we return null whenever we find out that an item has i.h + i.y > maxrows from there (similar to an old pr which focused on this)
and whenever compactedLayout is null - take the oldLayout (the valid one - which doesn't violate borders).

But sometimes even though the layout available in state is a valid one - it doesn't rerender.

We found a workaround to refresh/rerender is - change the compact type to some random value like uuid and then back to verticle.

pass this function as a prop to rgl and call this function everywhere compactedLayout null is handled - similar to previous step.

Now another thing popped up that even the invalid layout is stored in the state randomly. and now again everything goes back to where it was.

We also want to provide a pr once it's fixed.

Thanks in advance!
